src/main/java/com/openlattice/chronicle/services/edm/EdmCacheService.java

```diff
@@ -0,0 +1,242 @@
+                    .stream()
+                    .collect( Collectors.toMap( CollectionTemplateType::getName, AbstractSecurableObject::getId ) );
+
+            // for each config map orgId -> templateTypeName -> ESID
+            Map<UUID, Map<CollectionTemplateTypeName, UUID>> orgEntitySetMap = new HashMap<>();
+
+            configs.forEach( userAppConfig -> {
+                Map<UUID, UUID> templateTypeIdESIDMap = collectionsApi
+                        .getEntitySetCollection( userAppConfig.getEntitySetCollectionId() ).getTemplate();
+
+                // iterate over templateTypeName enums and create mapping templateTypeName -> entitySetId
+                Map<CollectionTemplateTypeName, UUID> templateTypeNameESIDMap = Maps
+                        .newHashMapWithExpectedSize( templateTypeIdESIDMap.size() );
+
+                for ( CollectionTemplateTypeName templateTypeName : CollectionTemplateTypeName.values() ) {
+                    UUID templateTypeId = templateTypeNameIdMap.get( templateTypeName.toString() );
+
+                    if ( templateTypeId != null ) {
+                        templateTypeNameESIDMap.put( templateTypeName, templateTypeIdESIDMap.get( templateTypeId ) );
+                    }
+                }
+                orgEntitySetMap.put( userAppConfig.getOrganizationId(), templateTypeNameESIDMap );
+            } );
+
+            entitySets.put( appComponent, orgEntitySetMap );
+        } );
+
+        entitySetIdsByOrgId.putAll( entitySets );
+    }
+
+    @Override
+    public UUID getPropertyTypeId( FullQualifiedName fqn ) {
+        return this.propertyTypeIdsByFQN.get( fqn );
+    }
+
+    @Override
+    public PropertyType getPropertyType( UUID propertyTypeId ) {
+        return propertyTypesById.get( propertyTypeId );
+    }
+
+    @Override
+    public PropertyType getPropertyType( FullQualifiedName fqn ) {
+        return getPropertyType( getPropertyTypeId( fqn ) );
+    }
+
+    @Override
+    public Map<String, UUID> getPropertyTypeIds( Set<String> propertyTypeFqns ) {
+        EdmApi edmApi;
+        try {
+            ApiClient apiClient = apiCacheManager.prodApiClientCache.get( ApiClient.class );
+            edmApi = apiClient.getEdmApi();
+        } catch ( ExecutionException e ) {
+            logger.error( "Unable to load EdmApi" );
+            return ImmutableMap.of();
+        }
+
+        return propertyTypeFqns.stream().map( FullQualifiedName::new ).map( fqn -> Pair
+                .of( fqn.getFullQualifiedNameAsString(),
+                        edmApi.getPropertyTypeId( fqn.getNamespace(), fqn.getName() ) ) )
+                .collect( Collectors.toMap( Pair::getLeft, Pair::getRight ) );
+    }
+
+    @Override
+    public Map<AppComponent, Map<UUID, Map<CollectionTemplateTypeName, UUID>>> getEntitySetIdsByOrgId() {
+        return entitySetIdsByOrgId;
+    }
+
+    @Override
+    public UUID getParticipantEntitySetId( UUID organizationId, UUID studyId ) {
```

Comment:
We are only hitting then EntitySetsApi when in legacy since we are not caching participant ESID for each study...

I think we could make some perf gains since this function is used in several places, including uploading data.